### PR TITLE
Pratster95 robot-name Solution

### PR DESCRIPTION
**Pull request by Pratster95**

### CODE REVIEW

File: robot-name/Pratster95/robot_name.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package robotname
+
+import (
+	"fmt"
+	"math/rand"
+)
+
+// Define the Robot type here.
+type Robot struct {
+	name string
```
Comment (davidmogar) [inline robot-name/Pratster95/robot_name.go:10]: Given that you are exporting the Robot struct (capitalized), it would make sense to expose its only field too.

File: robot-name/Pratster95/robot_name.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+var nameUsed = map[string]bool{}
+
+const max = 26 * 26 * 10 * 10 * 10
+
+func (r *Robot) Name() (string, error) {
+	if r.name != "" {
+		return r.name, nil
+	}
+	if len(nameUsed) >= max {
+		return "", fmt.Errorf("No Names left")
```
Comment (davidmogar) [inline robot-name/Pratster95/robot_name.go:22]: Errors in go should start with lowercase.
```suggestion
		return "", fmt.Errorf("no names left")
```

File: robot-name/Pratster95/robot_name.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	}
+	for {
+		name := getName()
+		if !nameUsed[name] {
```
Comment (davidmogar) [inline robot-name/Pratster95/robot_name.go:26]: I had no idea that the default value is returned for those non-existent keys. Nice.

I would still use the usual syntax for checking if the element is present though. It would be more consistent as it wouldn't change the condition depending in the values' type.

File: robot-name/Pratster95/robot_name.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+			return r.name, nil
+		}
+	}
+}
+
+func (r *Robot) Reset() {
+	r.name = ""
+}
+func getName() string {
+	return fmt.Sprintf("%v%v%v%v%v", string(rand.Int31n(26)+66), string(rand.Int31n(26)+66), rand.Int31n(10), rand.Int31n(10), rand.Int31n(10))
```
Comment (davidmogar) [inline robot-name/Pratster95/robot_name.go:38]: I'd say that `%v` is too generic and that is better to use `%s%s%d%d%d` in this case. Another change for this line would be to generate a single random number for the numerical part. So something like this:
```suggestion
	return fmt.Sprintf("%s%s%d", string(rand.Intn(26)+66), string(rand.Intn(26)+66), rand.Intn(1000))
```

File: robot-name/Pratster95/robot_name.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+type Robot struct {
+	name string
+}
+
+var nameUsed = map[string]bool{}
+
+const max = 26 * 26 * 10 * 10 * 10
+
+func (r *Robot) Name() (string, error) {
+	if r.name != "" {
```
Comment (davidmogar) [inline robot-name/Pratster95/robot_name.go:18]: I'm all in with proper names. I'd call it `robot` instead of `r`. It becomes easier to read the code and the compiler don't actually care about how many characters we use.

File: robot-name/Pratster95/robot_name.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	"fmt"
+	"math/rand"
+)
+
+// Define the Robot type here.
+type Robot struct {
+	name string
+}
+
+var nameUsed = map[string]bool{}
```
Comment (davidmogar) [inline robot-name/Pratster95/robot_name.go:13]: Correct me if I'm wrong, but the proper (most common) way of creating an empty map is:
```go
make(map[string]bool)
```
You don't need a literal declaration if you are not adding any key/value.

File: robot-name/Pratster95/robot_name.go (hunk 5 above)
Comment (davidmogar) [reply robot-name/Pratster95/robot_name.go:18]: **TLDR:** Forget my comment above.

It was brought to my attention that this could not be the best approach.

There's is no convention regarding this (that I could find) but the most common thing to find is one letter names like you used. Searching a bit more about this, I found out a wiki page in the go repository that collects common review comments. [It says this about receivers' names](https://github.com/golang/go/wiki/CodeReviewComments#receiver-names):

> The name of a method's receiver should be a reflection of its identity; often a one or two letter abbreviation of its type suffices (such as "c" or "cl" for "Client"). Don't use generic names such as "me", "this" or "self", identifiers typical of object-oriented languages that gives the method a special meaning. In Go, the receiver of a method is just another parameter and therefore, should be named accordingly. The name need not be as descriptive as that of a method argument, as its role is obvious and serves no documentary purpose. It can be very short as it will appear on almost every line of every method of the type; familiarity admits brevity. Be consistent, too: if you call the receiver "c" in one method, don't call it "cl" in another.

In short:

- One or two letters are more common for receivers' names.
- You should keep consistency.
- You can use whatever name you want but as it's role is well defined, you don't need to be descriptive.

Discussing about this, the only corner case which I could find is having more than one struct with receivers in the same go file. In those cases, the naming would become more confusing. Of course, declaring the receivers just after the struct would improve the readability.
